fire/cli/niv/_læs_observationer.py

```diff
@@ -261,66 +258,77 @@ def læs_observationsstrenge(
-                for line in obsfil:
-                    if "#" != line[0]:
-                        continue
-                    # Fjern luft i begge ender, havelågen i starten og kollaps gentagen luft
-                    line = re.sub(r"[ \t]+", " ", line.lstrip("# ").strip())
-
-                    # Check at observationen er i et af de kendte formater
-                    tokens = line.split(" ", 13)
-                    assert len(tokens) in (
-                        9,
-                        13,
-                        14,
-                    ), f"Deform input linje: {line} i fil: {fil.Filnavn}"
-
-                    # Bring observationen på kanonisk 14-feltform.
-                    for _ in range(len(tokens), 13):
-                        tokens.append(0)
-                    # Tilføj tom kommentar hvis der ikke er nogen med indhold
-                    if len(tokens) < 14:
-                        tokens.append('""')
-                    # Befri kommentar for anførelsestegn og overflødige mellemrum
-                    tokens[13] = tokens[13].lstrip('"').strip().rstrip('"')
-
-                    # Korriger de rædsomme dato/tidsformater
-                    tid = " ".join((tokens[2], tokens[3]))
-                    try:
-                        isotid = datetime.strptime(tid, "%d.%m.%Y %H.%M")
-                    except ValueError:
-                        sys.exit(
-                            f"Argh - ikke-understøttet datoformat: '{tid}' i fil: '{fil.Filnavn}'"
-                        )
-
-                    # Opbyg række-som-dict: Omsæt numeriske data fra strengrepræsentation til tal
-                    obs = {
-                        "Fra": tokens[0],
-                        "Til": tokens[1],
-                        "L": float(tokens[4]),
-                        "ΔH": float(tokens[5]),
-                        # Undgå journalside fortolkes som tal: Erstat decimalseparator
-                        "Journal": tokens[6].replace(".", ":"),
-                        "T": float(tokens[7]),
-                        "Opst": int(tokens[8]),
-                        "Sky": int(tokens[9]),
-                        "Sol": int(tokens[10]),
-                        "Vind": int(tokens[11]),
-                        "Sigt": int(tokens[12]),
-                        "σ": fil.σ,
-                        "δ": fil.δ,
-                        "Kommentar": tokens[13],
-                        "Sluk": "",
-                        "Hvornår": isotid,
-                        "Kilde": fil.Filnavn,
-                        "Type": fil.Type.upper(),
-                        "uuid": "",
-                    }
-                    observationer = observationer.append(obs, ignore_index=True)
+            # check for unicode læsefejl
+            obsfil.readlines()
+            # hvis ingen fejl mødes spoles filen tilbage til start
+            obsfil.seek(0)
+        except UnicodeDecodeError:
```

Comment:
Jeg er en smule loren ved om dette er det rette at gøre. Selvfølgelig er det det letteste, men fejlen opstår formodentlig fordi der et sted undervejs overføres filer med en syvbitsprotokol (a la ftp ascii, frem for ftp bin) - så der er FEJL i filen, enten fordi vi i sdfe behandler den forkert, eller fordi vores leverandør af måleinstrumentstyringsprogrammel har klokket i det.

Jeg *tror* jeg mener (men er villig til at lade mig overbevise om andet) at fejlen i stedet bør diagnosticeres og meddeles brugeren - meget gerne med et linjenummer for fejlen. Så kan den i første omgang rettes manuelt, og derigennem animere til at man får styr på alle filoverføringsprocesser i arbejdsgangene.

Reply:
Jeg er enig med dig i at det vil være bedst at rette problemet hvor det opstår men det er jeg ikke sikker på er så nemt. Hvis det er i instrumentprogrammel har vi ikke den store chance for at gøre noget men hvis ændret adfærd andre steder kan løse det er det værd at undersøge. Jeg har ingen anelse om hvordan arbejdsgangene er her. 

Så vidt jeg lige kan se er det et "å" der driller i den her fil. Formentligt er det noget lignende de fleste andre gange vi støder på den her fejl. Den er lidt svær at kommunikere en løsning på: "Åben filen, slet alle å'er og skriv nye samme sted". Det bliver hurtigt et irritationsmoment, især da der findes en måde at omgå det på. Det er i øvrigt ikke helt lige til at pege på den specifikke linje der har problemet - fejlen udløses i linjen med `for line in obsfil:`. Formentligt fordi den bagved kører `obsfil.readlines()` eller noget lignende. Så det vil fremstå som linje 0, selvom fejlen faktisk er på linje 3.

Det korte af det lange er, at jeg synes vi skal køre med den her løsning som godt nok ikke er perfekt, men dog fjerner noget friktion hos brugerne uden det har nogen reel ulempe (så vidt jeg kan se). Vi kan evt. skrive en advarsel på skærmen om filen er encodet forkert.

Reply:
En simplere løsning er vel i øvrigt bare at fjerne `encoding`-delen af `open()`-kaldet. Jeg antog at der var en speciel grund til at det blev gjort på den måde, men jeg ved faktisk ikke hvad den skulle være. Måske der er andre situationer hvor det går galt?

Reply:
Min pointe var at jeg tror at der er tale om at "nogen har glemt noget" i forbindelse med forgangne tiders arbejdsgange og flytter ting frem og tilbage med ftp (muligvis ubevidst, fordi det er den indbyggede klient i hvad-er-det-nu-det-tekstredigeringsprogram-hedder, der gør det).

Problemet sidste forår, hvis jeg husker korrekt (det var helt vanittigt),  viste sig som Unicodes "Ukendt Tegn"-codepoint, indkodet som ISO latin-1. Så det vistes korrekt, men i utf8 er det et to-byte-tegn, mens det i latin-1 er et 1-byte-tegn. Så et eller andet sted begås der vold mod filen - måske når den åbnes i et flinkt og hjælpsomt tekstredigeringsprogram, der er sat op til verden af 1985. Så jeg mener at det under alle omstændigheder skal elimineres.

Så min hovedpointe var at det er en **feature** at det bliver et irritationsmoment: Det er ikke en showstopper, man kan komme videre,  men man animeres til at finde fejlen (som sandsynligvis ender på et af vores borde under alle omstændigheder).

Jeg er ikke sikker på at det vil gå godt at dumpe en ugydig unicodesekvens i en utf-8 indkodet .xlsx-fil

Reply:
Har lige kigget lidt nærmere 

```
(fire-dev) C:\Users\B004330\Documents\2021\Projects>head -3 C:\Users\B004330\Downloads\obs_gl.txt | hd
00000000  50 72 6f 67 72 61 6d 3a  20 20 20 20 20 20 20 20  |Program:        |
00000010  20 20 20 20 4d 47 4c 20  31 2e 38 20 32 30 31 34  |    MGL 1.8 2014|
00000020  2d 30 36 2d 32 30 0d 0a  46 69 6c 6e 61 76 6e 3a  |-06-20..Filnavn:|
00000030  20 20 20 20 20 20 20 20  20 20 20 20 43 3a 5c 4e  |            C:\N|
00000040  69 76 50 72 6f 67 73 5c  72 65 73 75 6c 74 61 74  |ivProgs\resultat|
00000050  66 69 6c 65 72 2f 74 72  61 6e 73 70 5f 6f 64 64  |filer/transp_odd|
00000060  65 6e 5f 35 0d 0a 50 72  6f 6a 65 6b 74 62 65 73  |en_5..Projektbes|
00000070  6b 72 69 76 65 6c 73 65  3a 20 54 72 61 6e 73 70  |krivelse: Transp|
00000080  6f 6e 64 65 72 20 6f 67  20 72 65 66 6c 65 6b 74  |onder og reflekt|
00000090  6f 72 6d e5 6c 69 6e 67  20 76 65 64 20 48 61 76  |orm.ling ved Hav|
```
`0xe5` i fjerde tegnposition i nederste linje (den der starter med 00000090) er latin-1 for "å", dvs. codepoint `e*16+5` = 229: Se linje `&#229` [her](https://cs.stanford.edu/people/miles/iso8859.html), så i dette tilfælde er der bare tale om en latin-1 fil, som skal behandles som sådan.

Sidste forårs problem for mig var at der rent faktisk stod et UTF-8 "ukendt tegn glyph" codepoint i inputfilen - som derfor var gyldig unicode.

Hvis det viser sig at instrumentet skriver latin-1, så skal filerne "selvfølgelig" (for en tilstrækkeligt vidtgående definition af selvfølgelighed) "bare" (for en tilsvarende definition af *bare*) læses som latin-1, og under indlæsningen i Pythons automagik konverteres til UTF-8, så det passer ind i .xlsx. Der står noget om det [her](https://stackoverflow.com/questions/18260859/python3-ascii-utf-8-iso-8859-1-cant-decode-byte-0xe5-swedish-characters) helt nede i bunden af siden, som jeg ikke kan nå at læse grundigt lige nu, da jeg skal op og lave mad til min forslugne og hungrende familie :-)

Reply:
> Så min hovedpointe var at det er en feature at det bliver et irritationsmoment 

Så langt er jeg med. Jeg synes bare ikke der er vundet noget særligt hvis vi kan arbejde uden om problemet. Hvis det var dig og mig der blev udsat for den her fejl jævnligt ville vi hurtigt blive træt af den og finde en bedre løsning, men dem den rammer har kun muligheden for at blive irriteret. Løsningen kan de med stor sandsynlighed ikke finde frem til. Så er slutresultatet bare irritation. Hvis vi ved at gå lidt pragmatisk til værks kan undgå det synes jeg det er en success. Om jeg har fundet den rigtige løsning i dette PR ved jeg ikke, det er trods alt kun baseret på et enkelt tilfælde. 

> Jeg er ikke sikker på at det vil gå godt at dumpe en ugydig unicodesekvens i en utf-8 indkodet .xlsx-fil

Det ved jeg heller ikke, men vi finder helt sikkert ud af det hvis ikke det er muligt.

Reply:
> Løsningen kan de med stor sandsynlighed ikke finde frem til

Nej, det ville vi være nødt til at hjælpe dem med. Vi skal bare have en anledning til at gennemgå kommunikationsvejene

> Hvis vi ved at gå lidt pragmatisk til værks kan undgå det synes jeg det er en success

Det ville være et fint stop-gap, men hvis vi med lidt analyse kunne konstatere at inputfilerne *er* latin-1, og håndtere dem korrekt med den viden i baghånden, så er det da endnu bedre. Tegnsætsforbistringer er en pestilens, og et latin-1-tegn med bit 7 sat (som i æøåÆØÅ) dumpet i en UTF-8 er en fejl: Læseren af filen vil så antage at vi har fat i en flerbytesekvens, og samle videre op på næste byte i bytestrømmen. Når den så *ikke* har bit 7 sat, så kan læsekoden ikke gøre andet end at konstatere at filen er korrumperet.

UTF-8 er supersmart fundet på, men det er ikke kompatibelt med 1980ernes 8-bit tegnindkodninger ("code pages", "ISO 8859-x", latin-x, eller hvad de nu ellers er blevet kaldt undervejs...), så jeg

Reply:
> Det ville være et fint stop-gap

Ups, mangler en ledsætning her. Mente: "Det ville være et fint stop-gap, men det kræver udfiltrering af codepoints over 127"